### DISCUSSION TITLE
[unist-util-filter] Typescript: Property 'type' does not exist on type 'Test<Node>'

### DISCUSSION THREAD

**Discussion by bearcanrun** [syntax-tree/unist#29] [Q&A] [answered]
Attempting to type a utility that removes frontmatter from markdown. Functions works as just JS, I just can't figure out how to add 'type' property to `node`. Help, please! (disclaimer: I'm still learning TS)

```js
import filter from 'unist-util-filter'
import visit from 'unist-util-visit'
import type { Node } from 'unist'
import type { Test } from 'unist-util-is'

interface YamlNode extends Node {
  type: string
}

export const removeFrontmatter = () => async (tree: Node, file) => {
  let getFrontMatter
  await visit(tree, 'yaml', (node) => {
    getFrontMatter = node.data.parsedValue
    return
  })
  file.data.frontmatter = getFrontMatter
  return filter(tree, (node: Test<YamlNode>): node is YamlNode => node.type !== 'yaml')
// TS error here: Property 'type' does not exist on type 'Test<Node>' ^^^^^^^^
}
```

**Comment by ChristianMurphy** [comment, marked as answer]
A few different questions in here: 1 what is the input to a `Test` accept? 2 why is `type` missing? 3 what is a way to resolve this?

1. what is the input to a `Test` accept?
   ```js
   (node: Test<YamlNode>): node is YamlNode => node.type !== 'yaml'
   ```
   
   is known as a `Test` https://github.com/syntax-tree/unist-util-is#function-testnode-index-parent. A test may or may not receive a valid `node`, as such it doesn't know what it's receiving and `node` is [type unknown](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#new-unknown-top-type).
a `Test` can also be null which checks that a `node` is valid (but not it's type), `string` which check that a node is valid and that node type matches the string passed in, or an object which checks that it matches passed in attributes.
https://github.com/syntax-tree/unist-util-is#use

2. why is `type` missing? Because `node` is an unknown type, to access a property on an unknown type, it first has to be narrowed to guarantee that the property exists. https://fettblog.eu/typescript-hasownproperty

3. what is a way to resolve this? Since you mainly want to filter `node`s where type is `yaml`, using the `string` test type or `object` test type since they validate that `unknown` is a `Node` before doing the validation type/attribute check.
   ```js
   return filter(tree, 'yaml')
   ```

**Reply by bearcanrun** [reply]
Wow, super gratitude for spelling it out for me. I had reduced it down to your answer at some point, getting rid of the error, but I was really trying to understand the typing side of it. 

I had indeed followed the type definitions from [_filter_](https://github.com/syntax-tree/unist-util-filter#filtertree-options-test) param`filter?: Test<T>` &rarr; `type Test<T extends Node> = TestType<T> | TestObject<T> | TestFunction<T>` &rarr; `type TestType<T extends Node> = T['type']`, but generics still don't automatically translate for me and I hadn't comprehended the now obvious meaning of `T['type']`. Level up, thanks to you!
